### PR TITLE
59 rework taxonomy classifier

### LINKED ISSUE #59: Rework taxonomy classifier
- [x] Follow the same structure as the binary classifier
- [x] Use GPT-labelled data

### PR DESCRIPTION
**Pull request by RFOxbury**
---

# Description

This PR updates the pipeline for training the taxonomy classifier.

Fixes #59

# Instructions for Reviewer

In order to test the code in this PR you need to ...

Run the scripts in `discovery_child_development/pipeline/taxonomy_classifier/` in the following order:

- [ ] `01_train_test_split.py`
- [ ] `02a_sentence_embeddings.py`
- [ ] `02b_prep_hf_dataset.py` with `--production=False` (should run fine locally)
- [ ] `03_set_baseline.py`. You can try varying the parameters `model_type` and `wandb` although they have not changed since the last version of this script (`baseline_model.py`).
- [ ] `04b_finetune_distilbert.py` with `--production=False` and `--wandb=True`. Even when running it in test mode, it may still crash because (I think) the batch size is too much for our computers to handle? If this is the case, follow the instructions in `infra/skypilot/README.md` to ssh into a compute cluster and run the script there.

Please pay special attention to ...

- whether everything is suitably documented (I've probably missed something)!

# Checklist:

- [x] I have refactored my code out from `notebooks/`
- [x] I have checked the code runs
- [x] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [x] I have merged any new changes from `dev`
- [x] I have documented the code
  - [x] Major functions have docstrings
  - [x] Appropriate information has been added to `README`s
- [x] I have explained this PR above
- [x] I have requested a code review

### CODE REVIEW

File: discovery_child_development/getters/taxonomy.py
Comment (beingkk) [inline discovery_child_development/getters/taxonomy.py:1]: I see you've moved bunch of global variables from this file to taxonomy_classifier.py. I think you also ought to delete some of the functions from this file - I've done it now in another commit.

File: discovery_child_development/getters/taxonomy_classifier.py [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+    """
+
+    if split not in ["train", "test", "val"]:
+        raise ValueError(
+            f"Invalid value for 'split': {split}. Allowed values are 'train', 'test', 'val'."
+        )
+
+    s3_file = s3_path + s3_filename.replace("SPLIT", split)
+
+    return nesta_s3.download_obj(s3_bucket, s3_file, download_as="dataframe"), s3_file
```
Comment (beingkk) [inline discovery_child_development/getters/taxonomy_classifier.py:182]: Why do you output the `s3_file` as well - for checking?

File: discovery_child_development/pipeline/models/taxonomy_classifier/02a_sentence_embeddings.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+VECTORS_FILENAME = taxonomy_config["s3_vectors_name"]
+
+
+def embed_texts(
+    split="train",
+    s3_bucket=S3_BUCKET,
+    vectors_out_path=VECTORS_OUT_PATH,
+    vectors_filename=VECTORS_FILENAME,
+):
+    labelled_text_df = taxonomy_classifier.get_training_data(split=split)
```
Comment (beingkk) [inline discovery_child_development/pipeline/models/taxonomy_classifier/02a_sentence_embeddings.py]: This was throwing an error, so I changed to
```suggestion
    labelled_text_df, _ = taxonomy_classifier.get_training_data(split=split)
```

File: discovery_child_development/getters/taxonomy_classifier.py [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+# get_labelling_sample
+S3_LABELLING_DATA = (
+    "data/labels/taxonomy_classifier/training_validation_data_patents_openalex.jsonl"
+)
+LOCAL_PATH = PROJECT_DIR / "inputs/data/labelling/taxonomy/input"
+LOCAL_FILE = f"{LOCAL_PATH}/training_validation_data_patents_openalex.jsonl"
+create_directory_if_not_exists(LOCAL_PATH)
+
+# get_gpt_labelled_sample
+GPT_LABELLED_DATA = "data/labels/taxonomy_classifier/labelled_with_gpt/training_validation_data_patents_openalex_GPT_LABELLED.parquet"
```
Comment (beingkk) [inline discovery_child_development/getters/taxonomy_classifier.py:26]: Hey, would this be your previously labelled data or the newly generated using the "v2" pipeline?

**Review by beingkk — APPROVED** [review]
Hey Rosie, thanks a lot for this - I test ran the pipeline, including running it with skypilot, and it run alright! I ran into a some bugs, but I fixed those with the latest commit.

I guess this code is presently using the labels produced by your previous labelling pipeline, rather than the "v2" labels - and I'll need to adapt the code a bit accordingly?

I'll probably have more questions as I replicate the training process for the detection-management classifier. Hope that's OK, and I think it's good to merge.

Thank you!

File: discovery_child_development/getters/taxonomy.py
Comment (RFOxbury) [reply discovery_child_development/getters/taxonomy.py:1]: Ok, thank you Karlis! 🙏🏻

File: discovery_child_development/getters/taxonomy_classifier.py (hunk 1 above)
Comment (RFOxbury) [reply discovery_child_development/getters/taxonomy_classifier.py:182]: For logging a reference to the file on weights and biases (so you tell weights and biases what file was the input to the run)

File: discovery_child_development/pipeline/models/taxonomy_classifier/02a_sentence_embeddings.py (hunk 2 above)
Comment (RFOxbury) [reply discovery_child_development/pipeline/models/taxonomy_classifier/02a_sentence_embeddings.py]: 😶 thank you!
